README.md
<!-- This Source Code Form is subject to the terms of the Mozilla Public
   - License, v. 2.0. If a copy of the MPL was not distributed with this
   - file, You can obtain one at https://mozilla.org/MPL/2.0/. -->

The Spinitron Ark Player is a widget you can use on a web page. It allows
visitors to play recordings of a radio webcast from Spinitron's Ark
service.

## Demo and instructions

[Combined demo and instructions](https://spinitron.github.io/ark-player/) for web page authors on how to use the widget.

## Use the widget on your web site

You can use the widget on your web site if you comply with these [Terms of Service](https://forum.spinitron.com/t/ark-terms-of-service/277). The
demo page linked above has instructions.

## Access to audio files

If you run the Ark Player on a server of your own and for listeners to access archives managed by Spinitron, ask Spinitron to authorize your origin (CORS) to access the audio files.

## Using the widget as a dependency in your project

If you're using NPM then you can

```
npm install ark-player
```

and import it from there into your project.

## Developer instructions

If you intend to customize or change the player, e.g. to change its appearance
or behavior, please fork this repo in Github and push your changes
to your repo. (We use MPL for this work to encourage you to give your improvements
back to the community.) Feel free to send pull requests.

We use git, Sass, Babel, Rollup and NPM. Find instructions for using those elsewhere.

Clone the git repo locally and then, in the repo dir, run `npm install`.

After making your change in the `src` directory, `npm run build` should update the files in `dist`.

## Copyright and license

All files in this repository are Copyright Spinitron LLC.

You may use, modify, and/and redistribute this work under conditions of the
[Mozilla Public License 2.0](https://www.mozilla.org/en-US/MPL/).


changes.txt
4.3.2 2023 Jun 05

Maybe try to fix version confusion.


4.2.1 2023 Jun 05

Fixed looping bugs with hls.js v1.4.1 https://github.com/video-dev/hls.js/issues/5476
by removing recovery code, instead destroy the Hls player on fatal error.

Add version strings for ark-player and hls.js to user agent.

Fix client error (R2908) in case player UI doesn't exist.

Update station options in demo page dropdown.



webpack-libs.config.js
const path = require('path');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const TerserPlugin = require('terser-webpack-plugin');
const autoprefixer = require('autoprefixer');

const env = process.env.NODE_ENV === 'production' ? 'production' : 'development';

/**
 * Special webpack config to produce libraries.
 * https://webpack.js.org/guides/author-libraries/
 */
module.exports = {
    mode: env,
    devtool: false,
    entry: {
        arkPlayer: `./assets/scripts/arkPlayer.js`,
    },
    output: {
        path: path.resolve('web/static/lib'),
        filename: `Spinitron.[name].js`,
        library: ['Spinitron', '[name]'],
        libraryTarget: 'umd',
        umdNamedDefine: true,
    },
    resolve: {
        modules: [
            path.resolve('.'), // allow imports relative to the project root
            path.resolve('./node_modules'),
        ],
    },
    optimization: {
        nodeEnv: env,
        minimizer: [
            // JS minifier
            new TerserPlugin({ cache: true, parallel: false, sourceMap: false }),
            // CSS minifier
            new OptimizeCSSAssetsPlugin({
                cssProcessorPluginOptions: {
                    preset: ['default', { discardComments: { removeAll: true } }],
                },
                cssProcessorOptions: {
                    map: { inline: false, annotation: true },
                },
            }),
        ],
    },
    stats: {
        entrypoints: false,
        children: false,
        modules: false,
    },
    performance: {
        hints: false,
    },
    watchOptions: {
        aggregateTimeout: 2000,
        ignored: /node_modules/,
        // Use poll, since watching doesn't work with NFS and VirtualBox (i.e. with Vagrant too).
        poll: 3000,
    },
    module: {
        rules: [
            {
                // JS loader, we use babel-loader to transpile JS based on the babel config `babel.config.js`
                test: /\.js$/,
                exclude: /(node_modules|bower)/,
                use: {
                    loader: 'babel-loader',
                    options: { cacheDirectory: true },
                },
            },
            {
                // CSS loader, sass -> postcss to add vendor prefixes -> css -> extract to a file.
                test: /\.s?css$/,
                use: [
                    { loader: MiniCssExtractPlugin.loader },
                    { loader: 'css-loader', options: { sourceMap: false } },
                    {
                        loader: 'postcss-loader',
                        options: { ident: 'postcss', plugins: [autoprefixer()] },
                    },
                    { loader: 'resolve-url-loader', options: { keepQuery: true } },
                    { loader: 'sass-loader', options: { quiet: true } },
                ],
            },
            {
                test: /\.(png|svg|jpg|gif)$/,
                use: ['file-loader?name=../images/[name].[ext]'],
            },
        ],
    },
    plugins: [
        // CSS extractor
        new MiniCssExtractPlugin({
            filename: `../lib/[name].css`,
            chunkFilename: `../lib/[id].css`,
            ignoreOrder: true,
        }),
    ],
};


web/static/lib/Spinitron.arkPlayer.js
!function(e,t){"object"==typeof exports&&"object"==typeof module?module.exports=t():"function"==typeof define&&define.amd?define("arkPlayer",[],t):"object"==typeof exports?exports.arkPlayer=t():(e.Spinitron=e.Spinitron||{},e.Spinitron.arkPlayer=t())}(window,(function(){return function(e){var t={};function n(r){if(t[r])return t[r].exports;var o=t[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,n),o.l=!0,o.exports}return n.m=e,n.c=t,n.d=function(e,t,r){n.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},n.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},n.t=function(e,t){if(1&t&&(e=n(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(n.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var o in e)n.d(r,o,function(t){return e[t]}.bind(null,o));return r},n.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return n.d(t,"a",t),t},n.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},n.p="",n(n.s=2)}([function(e,t,n){},function(e,t,n){e.exports=n.p+"../images/loudspeaker.svg"},function(e,t,n){"use strict";n.r(t);n(0),n(1);const r=window.AudioContext||window.webkitAudioContext?new(window.AudioContext||window.webkitAudioContext):null,o=300,a=15;window.arkPlayer=function(e,t){function n(){e.outerHTML="<div>The archive player doesn't work in this browser :(</div>",document.querySelectorAll(".ark-play-button").forEach(e=>e.classList.add("ark-ark-play-button_hide"))}try{if("object"!=typeof r||"function"!=typeof r.resume||"string"!=typeof r.state)return void n()}catch(e){return void n()}const{template:i,data:s={},timeZone:u,enableDebug:l=!1,segmentDuration:c=o,fetchAhead:d=a}=t;let f=e.dataset.arkStart;const p=Object.keys(s).reduce((e,t)=>e.replaceAll(`{${t}}`,s[t]),i);let m=d;const g={dates:[],hours:{}};I("ark player init: ",t,"baseUrl",p);const y=(()=>{let e="?";try{e=(new Date).toLocaleString(["en-US",{timeZone:u}])}catch(e){return[]}return I(e),["en-US",{timeZone:u}]})(),h=e.querySelector(".ark-player__control"),v=e.querySelector(".ark-player__status"),S=e.querySelector(".ark-player__picker");let T,k=null,w=null,b=null,_=null,L=5;function x(t){const n=void 0===t?f:t,r=n.slice(0,11);let o;if(g.dates.some(e=>g.hours[e].some(t=>r===t[1]&&(o=e,!0))),o){const t=e;t.querySelector("[name=date]").value=o,N(t.querySelector("[name=hours]"),o),t.querySelector("[name=hours]").value=r,t.querySelector("[name=minutes]").value=(""+15*Math.floor(parseInt(n.slice(11,13),10)/15)).padStart(2,"0")}}function q(){"not-started"===M()?f&&D(f):function(e){const t=M(),n=void 0===e?"playing"!==t:e;"playing"!==t||n?"paused"===t&&n&&j("running"):j("suspended");A(n)}()}O(),(()=>{const t=new Date,n=t.getTime()-12096e5;let r,o,a=new Date(t.getFullYear(),t.getMonth(),t.getDate(),t.getHours()).getTime();for(;a>n;)a-=36e5,r=new Date(a),o=r.toLocaleDateString(...y).replace(/\d\d(\d\d)$/,"$1"),g.dates.includes(o)||(g.dates.push(o),g.hours[o]=[]),g.hours[o].push([r.toLocaleTimeString(...y).replace(/:00:00 /," "),r.toISOString().substring(0,13).replace(/-/g,"")]);const i=e.querySelector("[name=date]"),s=e.querySelector("[name=hours]"),u=e.querySelector("[name=minutes]");var l;function c(){O(),f=`${s.value}${u.value}00Z`}l=i,g.dates.forEach((e,t)=>{l[t]=new Option(e,e)}),i.addEventListener("change",()=>{N(s,i.value),c()}),s.addEventListener("change",c),N(s,i.value),u.addEventListener("change",c),function(e){["00","15","30","45"].forEach((t,n)=>{e[n]=new Option(":"+t,t)})}(u),x()})();const $=(()=>{let e=!1;const t=document.querySelector(".ark-player__volume-slider"),n=document.querySelector(".ark-player__volume-slash"),o=document.querySelector(".ark-player__mute-button");let a;const i={muteButton:()=>{e=!e,i.either()},slider:()=>{a=parseInt(t.value,10),i.either()},either:()=>{t.value=e?"0":""+a,n.style.cssText=e||0===a?"":"display:none";const o=(e?0:a)/100;b&&b.gainNode&&b.gainNode.gain.setValueAtTime(o,r.currentTime),_&&_.gainNode&&_.gainNode.gain.setValueAtTime(o,r.currentTime)}};return t.addEventListener("change",i.slider),t.addEventListener("input",i.slider),o.addEventListener("click",i.muteButton),i.slider(),i})();function D(e){if(!window.AudioContext&&!window.webkitAudioContext)return;O(!1),x(e);const t=e.match(/^(\d\d\d\d)(\d\d)(\d\d)T(\d\d)(\d\d)(\d\d)Z$/).slice(1).map(e=>parseInt(e,10));w=Date.UTC(t[0],t[1]-1,t[2],t[3],t[4],t[5]);let n=Date.UTC(t[0],t[1]-1,t[2],t[3],0,0);for(;n+1e3*c<=w;)n+=1e3*c;j("running"),P({segmentTime:n,offset:(w-n)/1e3}),T=setInterval(E,1e3)}function O(e){T&&clearInterval(T),_&&_.source&&_.source.stop(),_=null,b&&b.source&&b.source.stop(),b=null,!1!==e&&j("suspended"),w=null,k=null,v.querySelector(".ark-player__date").innerHTML="--/--/--",v.querySelector(".ark-player__time").innerHTML="-:--:-- --",v.classList.toggle("waiting",!0)}function M(){return null===b&&null===_?"not-started":"running"===r.state?"playing":"suspended"===r.state?"paused":(I(`player state?! current: ${b} next: ${_} context: ${r}. resetting`),O(),"not-started")}function j(e){if(r.state!==e){if("suspended"===e)r.suspend();else{if("running"!==e)return;r.resume()}!function t(){r.state!==e&&setTimeout(t,3)}()}}function A(t){const n=void 0===t?"playing"===M():t,r="ark-player_hide";e.classList.toggle(r,!1),h.querySelector(".ark-player__play").classList.toggle(r,n),h.querySelector(".ark-player__pause").classList.toggle(r,!n),S.classList.toggle(r,n),v.classList.toggle(r,!n)}function E(){if("paused"!==M()){if(b&&null===_){const e=b.started+b.source.buffer.duration-b.offset;r.currentTime>e-m&&P({segmentTime:b.segmentTime+1e3*c,offset:0})}if(_&&_.source&&_.started<r.currentTime&&(b=null,b=_,_=null,I("moved next segment to current")),k&&r.currentTime>k){const e=new Date(w+Math.round(1e3*(r.currentTime-k)));v.querySelector(".ark-player__date").innerHTML=e.toLocaleDateString(...y).replace(/(^\d\d?\/\d\d?\/)\d\d(\d\d)/,"$1$2"),v.querySelector(".ark-player__time").innerHTML=e.toLocaleTimeString(...y),v.classList.toggle("waiting",!1)}}}h.addEventListener("click",q),document.addEventListener("click",t=>{const n=t.target.closest("[data-ark-start]");n&&n!==e&&(t.preventDefault(),D(n.dataset.arkStart))}),document.addEventListener("keydown",e=>{e.isComposing||229===e.keyCode||e.altKey||e.ctrlKey||e.metaKey||e.shiftKey||("k"===e.key?q():"m"===e.key&&$.muteButton())});const C=(()=>{const e=[];return{add:({url:t,audio:n})=>{if(e.push({url:t,audio:n}),I("audio cache push "+t),e.length>3){const t=e.shift();I("audio cache discard "+t.url),t.audio=null}I("audio cache content: ",e)},get:t=>{for(let n=0;n<e.length;n+=1)if(e[n].url===t)return I("audio cache hit: "+t),e[n].audio;return I("audio cache miss: "+t),null}}})();function P({segmentTime:e,offset:t}){const n=p.replace("{time}",new Date(e).toISOString().replace(/[-:]/g,"").replace(/\.\d{3}/,""));I(`cuing next segment ${n} offset: ${t}`);const o=C.get(n);if(null!==o)return void H({segmentTime:e,offset:t},o);const a=new XMLHttpRequest,i=(new Date).getTime();a.open("get",n,!0),a.responseType="arraybuffer",a.onerror=()=>{I(`${a.statusText} getting ${n}`)},a.onload=()=>{200===a.status?(L=5,r.decodeAudioData(a.response,r=>{C.add({url:n,audio:r});const o=((new Date).getTime()-i)/1e3;I(`fetch and decode took ${o} s`),o+5>m&&(m=o+5,I(`increase fetch ahead time to ${m} s`)),H({segmentTime:e,offset:t},r)})):L>0?(I("Skip that segment and move on"),L-=1,P({segmentTime:e+1e3*c,offset:0})):(I("Failed to fetch audio after several tries. Resetting player"),O())},_=a,a.send(),A(!0)}function H({segmentTime:e,offset:t},n){const o=r.createGain();o.connect(r.destination);const a=r.createBufferSource();a.buffer=n,a.connect(o);const i=null===b?r.currentTime+.05:Math.max(r.currentTime+.01,b.started+b.source.buffer.duration-b.offset);_={segmentTime:e,source:a,offset:t,started:i,gainNode:o},$.slider(),_.source.start(i,_.offset),null===b&&(k=i),A(),I(`scheduled next segment to start at ${i} offset: ${t}`)}function I(...e){l&&console.log(`${(new Date).toLocaleTimeString()} ac.ct=${r.currentTime}`,...e)}function N(e,t){const n=e,r=n.selectedOptions[0]&&n.selectedOptions[0].innerText;for(;n.options.length>0;)n.remove(0);const o=g.hours[t].length-1;for(let e=0;e<=o;e+=1)n[e]=new Option(g.hours[t][o-e][0],g.hours[t][o-e][1]),r&&n[e].innerText===r&&(n[e].selected=!0)}}}])}));

src/arkPlayer.js
/* Copyright Spinitron LLC */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

window.ark2Player = function (container, options) {
    const {
        stationName,
        timeZone,
        enableDebug = false,
        hlsBaseUrl = 'https://ark2.spinitron.com/ark2',
        errorReportUrl = 'https://ark2.spinitron.com/errorlog/',
    } = options;

    /** @type {HTMLMediaElement} */
    var theMediaElement = container.querySelector('.ark-player__media-element');
    var myhls;
    var pleaseReload = true;
    let baseStart =
        container.dataset.arkStart ||
        `${new Date(new Date().getTime() - 7200000).toISOString().substring(0, 13).replace(/-/g, '')}0000Z`;
    const pickerOpts = { dates: [], hours: {} };
    var arkStartTime = null;
    const userAgent = window.navigator.userAgent + ' ark-player@v4.3.2 hls.js@' + Hls.version;

    debug('ark player init: ', options);

    // Uee localeStuff as argument to Date.toLocaleString() etc. If browser can't do timezones, empty array
    // means use client's time zone instead of the station's.
    const localeStuff = (() => {
        let β = '?';
        try {
            β = new Date().toLocaleString(['en-US', { timeZone }]);
        } catch (ignore) {
            return [];
        }
        debug(β);
        return ['en-US', { timeZone }];
    })();

    const playerControl = container.querySelector('.ark-player__control');
    const playerStatus = container.querySelector('.ark-player__status');
    const playerPicker = container.querySelector('.ark-player__picker');

    resetPlayer();

    // Initialize pickerOpts, the date-time picker database.
    (() => {
        const n = new Date();
        const stop = n.getTime() - 14 * 86400 * 1000;
        let hour = new Date(n.getFullYear(), n.getMonth(), n.getDate(), n.getHours()).getTime();
        let hourDate;
        let date;
        while (hour > stop) {
            hour -= 3600 * 1000;
            hourDate = new Date(hour);
            date = hourDate.toLocaleDateString(...localeStuff).replace(/\d\d(\d\d)$/, '$1');
            if (!pickerOpts.dates.includes(date)) {
                pickerOpts.dates.push(date);
                pickerOpts.hours[date] = [];
            }
            pickerOpts.hours[date].push([
                hourDate.toLocaleTimeString(...localeStuff).replace(/:00:00 /, ' '),
                hourDate.toISOString().substring(0, 13).replace(/-/g, ''),
            ]);
        }

        // Initialize the date-time picker selectors and their options.
        const dateEl = container.querySelector('[name=date]');
        const hoursEl = container.querySelector('[name=hours]');
        const minutesEl = container.querySelector('[name=minutes]');

        populateDate(dateEl);
        dateEl.addEventListener('change', () => {
            populateHours(hoursEl, dateEl.value);
            pickerChanged();
        });

        hoursEl.addEventListener('change', pickerChanged);
        populateHours(hoursEl, dateEl.value);

        minutesEl.addEventListener('change', pickerChanged);
        populateMinutes(minutesEl, 15);

        function pickerChanged() {
            resetPlayer();
            baseStart = `${hoursEl.value}${minutesEl.value}00Z`;
        }

        setupPicker();
    })();

    /**
     * Adjust the date-time picker to the given start.
     * @param {String} timestamp in ark player format
     */
    function setupPicker(timestamp) {
        const copyOfTimestampParam = timestamp === undefined ? baseStart : timestamp;
        const seekDateHour = copyOfTimestampParam.slice(0, 11);

        // Search the date-time picker's database for seekDateHour
        let date;
        pickerOpts.dates.some((d) => {
            return pickerOpts.hours[d].some((h) => {
                if (seekDateHour === h[1]) {
                    date = d;
                    return true;
                }
                return false;
            });
        });

        if (date) {
            // Adjust the picker's selections and hour options.
            const c = container;
            c.querySelector('[name=date]').value = date;
            populateHours(c.querySelector('[name=hours]'), date);
            c.querySelector('[name=hours]').value = seekDateHour;
            c.querySelector('[name=minutes]').value = `${
                Math.floor(parseInt(copyOfTimestampParam.slice(11, 13), 10) / 15) * 15
            }`.padStart(2, '0');
        }
    }

    function resetPlayer() {
        debug('resetPlayer');
        arkStartTime = null;
        pleaseReload = true;
        playerStatus.querySelector('.ark-player__date').innerHTML = '--/--/--';
        playerStatus.querySelector('.ark-player__time').innerHTML = '-:--:-- --';
        playerStatus.classList.toggle('waiting', true);
    }

    function playButton() {
        if (pleaseReload) {
            startArk(baseStart);
        } else {
            if (theMediaElement.paused) {
                theMediaElement.play();
            } else {
                theMediaElement.pause();
            }

            updateDisplay();
        }
    }

    function pressPlay() {
        theMediaElement.play();
        theMediaElement.ontimeupdate = () => {
            updateTime();
        };
        updateDisplay();
    }

    function startArk(startTimestamp) {
        debug(`startArk(${startTimestamp})`);

        setupPicker(startTimestamp);

        const d = startTimestamp
            .match(/^(\d\d\d\d)(\d\d)(\d\d)T(\d\d)(\d\d)(\d\d)Z$/)
            .slice(1)
            .map((δ) => parseInt(δ, 10));
        arkStartTime = Date.UTC(d[0], d[1] - 1, d[2], d[3], d[4], d[5]);

        const playlistUrl = `${hlsBaseUrl}/${stationName}-${startTimestamp}/index.m3u8`;

        if (Hls.isSupported()) {
            debug("Using Hls.js");
            if (myhls === undefined) {
                myhls = new Hls({
                    debug: enableDebug,
                    xhrSetup: function (xhr, url) {
                        xhr.setRequestHeader('User-Agent', userAgent);
                    },
                    fetchSetup: function (context, initParams) {
                        initParams.headers = new Headers({'User-Agent': userAgent});
                        return new Request(context.url, initParams);
                    },
                });
            }
            myhls.loadSource(playlistUrl);
            myhls.attachMedia(theMediaElement);
            myhls.on(Hls.Events.MEDIA_ATTACHED, function () {
                pleaseReload = false;
            });
            myhls.on(Hls.Events.ERROR, handleHlsError);
            myhls.on(Hls.Events.MANIFEST_PARSED, function (event, data) {
                pressPlay();
            });
            document.querySelector('.ark-player__mode-hint').innerHTML = 'H';
        } else if (theMediaElement.canPlayType('application/vnd.apple.mpegurl')) {
            debug("Using native HLS");
            theMediaElement.src = playlistUrl;
            pleaseReload = false;
            pressPlay();
            document.querySelector('.ark-player__mode-hint').innerHTML = 'N';
        }
    }

    const volume = (() => {
        let muted = false;
        const slider = document.querySelector('.ark-player__volume-slider');
        const slash = document.querySelector('.ark-player__volume-slash');
        const mute = document.querySelector('.ark-player__mute-button');
        let vol;
        const m = {
            muteButton: () => {
                muted = !muted;
                m.either();
            },
            slider: () => {
                vol = parseInt(slider.value, 10);
                m.either();
            },
            either: () => {
                slider.value = muted ? '0' : `${vol}`;
                slash.style.cssText = muted || vol === 0 ? '' : 'display:none';
                theMediaElement.volume = (muted ? 0 : vol) / 100;
            },
        };
        slider.addEventListener('change', m.slider);
        slider.addEventListener('input', m.slider);
        mute.addEventListener('click', m.muteButton);
        m.slider();

        return m;
    })();

    /**
     * Play/pause button.
     */
    playerControl.addEventListener('click', playButton);

    /**
     * Start playing when almost any element with specified [data-ark-start] is clicked.
     */
    document.addEventListener('click', (e) => {
        const element = e.target.closest('[data-ark-start]');
        if (!element || element === container) {
            return;
        }
        e.preventDefault();
        startArk(element.dataset.arkStart);
    });
    document.addEventListener('keydown', (e) => {
        if ((document.activeElement && document.activeElement.name && document.activeElement.name === 'message')
            || e.isComposing || e.keyCode === 229 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey) {
            return;
        }
        if (e.key === 'k') {
            playButton();
        } else if (e.key === 'm') {
            volume.muteButton();
        }
    });

    function updateDisplay() {
        const hide = 'ark-player_hide';
        // Show the player if it was hidden when no start time provided.
        container.classList.toggle(hide, false);
        playerControl.querySelector('.ark-player__play').classList.toggle(hide, !theMediaElement.paused);
        playerControl.querySelector('.ark-player__pause').classList.toggle(hide, theMediaElement.paused);
        playerPicker.classList.toggle(hide, !theMediaElement.paused);
        playerStatus.classList.toggle(hide, theMediaElement.paused);
    }

    function updateTime() {
        if (arkStartTime) {
            const arkTime = new Date(arkStartTime + Math.round(1000 * theMediaElement.currentTime));
            const timestring = arkTime
                .toLocaleDateString(...localeStuff)
                .replace(/(^\d\d?\/\d\d?\/)\d\d(\d\d)/, '$1$2');
            // Test the player elements still exist. R2908
            const playerDate = playerStatus.querySelector('.ark-player__date');
            if (playerDate) {
                playerDate.innerHTML = timestring;
            }
            const playerTime = playerStatus.querySelector('.ark-player__time');
            if (playerTime) {
                playerTime.innerHTML = arkTime.toLocaleTimeString(...localeStuff);
            }
            playerStatus.classList.toggle('waiting', false);
        }
    }

    function telemetry(data) {
        var xhr = new XMLHttpRequest();
        xhr.open('POST', errorReportUrl, true);
        xhr.setRequestHeader('Content-Type', 'application/json');
        const body = JSON.stringify({
            userAgent: navigator.userAgent,
            location: document.location.href,
            stationName,
            timeZone,
            hlsBaseUrl,
            data,
        });
        xhr.send(body);
    }

    function debug(...stuff) {
        if (enableDebug) {
            console.log(`${new Date().toLocaleTimeString()} `, ...stuff);
            //telemetry(stuff);
        }
    }

    function handleHlsError(event, data) {
        if (data.fatal) {
            //telemetry(data);

            // Follow instructions from Rob Walch https://github.com/video-dev/hls.js/issues/5476
            // Fatal error means all hls.js attempts to recover have failed.
            // Destroy the player ...
            myhls.destroy();
            // TODO
            // ...and create a new instance with a startPosition matching video.currentTime
            // or the end of the fragment that errored or last FRAG_CHANGED fragment.
        }
    }

    /**
     * @param {HTMLElement} selectEl
     */
    function populateDate(selectEl) {
        pickerOpts.dates.forEach((date, i) => {
            // eslint-disable-next-line no-param-reassign
            selectEl[i] = new Option(date, date);
        });
    }

    /**
     * @param {HTMLElement} selectEl
     */
    function populateHours(selectEl, date) {
        const hourSelect = selectEl;
        const previousSelection = hourSelect.selectedOptions[0] && hourSelect.selectedOptions[0].innerText;
        while (hourSelect.options.length > 0) {
            hourSelect.remove(0);
        }

        const l = pickerOpts.hours[date].length - 1;
        for (let i = 0; i <= l; i += 1) {
            hourSelect[i] = new Option(pickerOpts.hours[date][l - i][0], pickerOpts.hours[date][l - i][1]);
            if (previousSelection && hourSelect[i].innerText === previousSelection) {
                hourSelect[i].selected = true;
            }
        }
    }

    /**
     * @param {HTMLElement} selectEl
     * @param {number} step
     */
    function populateMinutes(selectEl) {
        ['00', '15', '30', '45'].forEach((minute, i) => {
            // eslint-disable-next-line no-param-reassign
            selectEl[i] = new Option(`:${minute}`, minute);
        });
    }
};
